### ISSUE TITLE
Konstant dice that contribute negative values should also contribute negative fire values

### ISSUE THREAD

**Issue by blackshadowshade** [buttonmen-dev/buttonmen#1802]
In #1717, Chaos found that konstant dice that are contributing negative values contribute *positive* fire values when another die with Fire is turned down. This is a problem that occurs with the buttons Hawaii and Arizona, as well as RandomBMMixed.

There is also a responder test in #1717.

**Comment by cgolubi1** [comment]
I'm going to argue that this is a blocker for merging #1861 --- right now, there are no named buttons containing both Konstant and Fire, but #1861 will enable Arizona and Hawaii.

**Comment by blackshadowshade** [comment]
I'm working on this at the moment. Just thought that I'd mention to @cgolubi1 that I've created a responder test for this (based on a previous responder test involving Hawaii), so you won't need to worry about that.

**Comment by jl8e** [comment]
I wish you well. I know I tried to figure out how to fix it without a rewrite of the fire logic and got nowhere. Definitely a mistake on my part way back when.

**Comment by jl8e** [comment]
Of course, right after posting that, my brain went "But what about...."

I have thoughts if you want them.

**Comment by blackshadowshade** [comment]
Thoughts are always welcome, Julian!  👍 

I worked this out before, but didn't write it down, and so I've forgotten again. I'll work it through again. However, it's something to do with splitting the logic out in one of the functions related to help values, I believe.

**Comment by jl8e** [comment]
There's probably several ways to approach it.

The way I was looking at was mostly in collect_firing_maxima. If, instead of a specific number, each die gave it a array of [point cost, current value, new value] entries, you could then test fire options by subtracting current_value and adding new_value.

(so, a normal die would return [ [1, 5, 6], [2, 5, 7], [3, 5, 8]], and a konstant would return [[1, 5, 6], [1, -5, -6], [2,5,7]...])

Since it's only skill attacks that care, there could be an additional method on each die to return the complex firing maxima info, and have skill attacks override the normal collect_firing_maxima to use it, which isolates all the changes to the attack code.

**Comment by blackshadowshade** [comment]
My approach was going to be directly altering BMAttackSkill->help_bounds() to produce an array of possible help values, which would not necessarily be contiguous.

These help values would be calculated with the nuance that konstant can contribute either negative or positive, probably not using hooks in a first iteration. If I get the logic working, I can always farm the relevant bit out to the die skill using hooks later.

**Comment by blackshadowshade** [comment]
I'll just note here that there's going to be a conflict here once #2389 is merged, since I put in an optimisation there to only calculate the hit table up to a reasonable value, as opposed to the entire hit table.

**Comment by blackshadowshade** [comment]
I've also stumbled across #1803.

We don't actually specify which dice are turned up when a fire attack is used, which means that the implicit implementation is that fire dice don't actually change the value of the other attacking dice.

Unfortunately, this is an issue for konstant dice, which actually would retain their new value.

My new calculation will determine what the value of the konstant die would need to be turned up to, as long as there is only one konstant attacker present.

At present, the only button that has multiple konstant dice is Nala, which does not also have Fire: Mkw(1) Mk(1) Mk(3) Mk(9) Mk(27)

Maybe it'll be enough to ban buttons in the future that have Fire and multiple Konstant dice.

**Comment by blackshadowshade** [comment]
And then, that leaves the ambiguity of what happens when konstant dice contribute a positive amount to a skill attack. In such a case, the fire support can often be distributed in a number of different ways across the attackers, meaning that the final value of the konstant die cannot necessarily be determined from the amount of fire support used.

**Comment by jl8e** [comment]
That would require user interaction, which would be a pain

Pragmatic solution: get fire-konstant interaction working. Mark the inability to adjust konstant dice as a bug, and work on it later. While it's a related issue, it's not the same issue, and it's easier. (It can be fed from the code that makes the user turn down their fire dice.)

(Also, it's going to need to be able to handle firing stinger konstant dice, where you can potentially set the die to a value you choose,)

**Comment by blackshadowshade** [comment]
Let me drill down into that a little bit. When you say get the fire-konstant interaction working, do you mean that I should assume that fire support is by default distributed to other dice than the konstant die, unless it's absolutely necessary to change the konstant die value?

**Comment by blackshadowshade** [comment]
So, we've had the suggestion from the forums (http://www.buttonweavers.com/ui/forum.html#!threadId=538) that perhaps Fire dice should not be able to change the value of konstant dice at all via fire support.

What do you think of that idea?

**Comment by irilyth** [comment]
I said this on the forum post, but I'll say it here too: I don't fundamentally object to that, but this would be a rule change to Konstant dice, not just a matter of interpretation -- the current rule is very clear that Fire (and Focus) can change the value of a Konstant die. But again, if we want to change that rule, I think that's basically ok.

**Comment by jl8e** [comment]
What I meant was making negative konstant and fire work, and worry about permanently adjusting the konstant die later. I was going to say to call it a bug, but you can always wontfix it later.

I think it's better to implement (or not) the die adjustment in a separate pass.

**Comment by TheOrgg** [comment]
The original rules to Konstant don't favor the Help Page rules in this case.  Having a Konstant remain Konstant in the face of fire seems to be the proper behavior.

If someone else would like to contact the creator of the Konstant die via Board Game Geek, maybe he could help make the ruling on this?

https://boardgamegeek.com/user/jkaufeld

**Comment by jl8e** [comment]
I am not an advocate of asking the designers of individual skills how they interact with other skills. The honest answer is pretty much always going to be "I never thought about it", and then an answer based on a relatively brief amount of thought. They don't necessarily know now what they were thinking then.

Like it or not, we are the experts here. Except for perhaps James Ernest,  almost nobody else has the background in thinking about skill interactions.

The factors to consider:

- The designer''s intent, based on the rules as written, understanding that they were written without consideration of other skills
- considerations of gameplay. (For instance, an interpretation of a skill that would require the non-active player to make decisions in the middle of their opponent's turn would lead to bad gameplay)
- considerations of implementation 
- considerations of button designer intent (if a button has a kF die, they presumably intend both skills to work. If they don't, is it a bug, or do we mean it?)
- considerations of consistency  (if a kF die can roll down, surely a fire die can roll a konstant die up)

So, having said all that, I now think that:
- kF dice should work (as I think they do). 
- Therefore, fire dice should be able to affect other konstant dice
- The definition of fire says that boosting a konstant die would change its actual value
- This is a hairy bit of implementation, and I'm not doing it,  so if fire can't affect konstant dice, I won't complain. If fire can affect konstant,  not changing the value is a bug, but a low-priority one

**Comment by TheOrgg** [comment]
I just broke out the original rules for Fire to see if there was a solution here.  To quote the last line of rules on Fire Dice:  "Dice can never be increased or decreased outside their normal range, I. E., a ten sided die can never show a number lower than 1 or larger than 10.  Also Fire Dice cannot assist other unusual dice in making attacks other than normal skill and power attacks."

So is using a value as a negative number a 'normal' skill attack?  I'd argue no, and thus Fire cannot be used as a negative number when contributing to a Konstant die's skill attack.

Now the adjustment of the Konstant die's value is the issue, not the negative Fire contribution.

**Comment by blackshadowshade** [comment]
That's an interesting solution. Do we say that Fire cannot assist a negative skill attack? If so, I may be able to program that without too much grief.

**Comment by TheOrgg** [comment]
I knew I had the set and while researching for the wiki, the buttonweavers blog came up.  I decided to see if Fire might say something similar to it 'adding to the attack' making this a moot point, but it instead said as above.
![IMG_20230212_131439](https://user-images.githubusercontent.com/44233993/218332310-6924076f-ef68-46ac-8d73-ebb4a4ddd9dc.jpg)
(pic, with backwards-upside-down Adam Spam blurb!  Also, just noticed we don't have the Polycon flavor text!  On it!)

**Comment by TheOrgg** [comment]
So does a kF die work as is?

**Comment by jl8e** [comment]
Counterargument: it's not adjusting the die to a negative value. It's adjusting the die to a value in its range, which can be used as either a positive or negative value.

It's pretty clear (IMO) that it would work -- before making the skill attack, you turn down the fire die, and turn up the konstant, then you make the attack.

**Comment by TheOrgg** [comment]
So your argument is that using Konstant as a negative number in a skill attack is still a normal skill attack, and that a Konstant die isn't an 'unusual' die type?

**Comment by jl8e** [comment]
If using konstant makes it not a normal skill attack, so does stinger.

That entire sentence is just "fire dice can't be used to aid weird attack types". It's redundant with the actual description of the mechanics. Which is not necessarily bad, but you're trying to read too much into it to justify not supporting the fire-konstant interaction.

But we don't need the excuse; if we want to not support it, we can just not support it. (My opinions on that question are written above, and have not changed. I also think that we can support it cleanly, but it will come with a rewrite of some parts of the code that need the work for other reasons.)

**Comment by TheOrgg** [comment]
Frankly, I would like to see the Fk interaction as outlined above and assumed; it gives a greater depth to the gameplay.  However, if sticking to the rules of the skill is important, I provided the rule.

This is why I've spent literal days worth of hours trying to track down the rules text for Gamer Dice.

**Comment by irilyth** [comment]
> Fire Dice cannot assist other unusual dice in making attacks other than normal skill and power attacks.

This sounds to me like it's referring to things like assisting Speed dice in making Speed Atatcks. I agree with @jl8e that they don't refer to things like assisting Stinger dice in making Skill Attacks; a Stinger die is "an unusual die", but a Skill Attack that you make with a Stinger die is still a normal Skill Attack. It's just that the Stinger die can contribute less to the attack than the number showing on it.

Konstant seems very similar to me: Where a `g(8):5` can contribute 1, 2, 3, 4, or 5 to a Skill Attack, a `k(8):5` can contribute 5 or -5 to a Skill Attack. Neither case turns the Skill Attack into something other than "a normal Skill Attack".

**Comment by PoshFrosh** [comment]
I agree with the other here who say that using a k die in a skill attack (to add _or_ subtract) is just a "normal" skill attack (like referenced in the Fire skill text) and not an "unusal" attack type.
This is supported by the current behavior on staging (as long as that behavior is intended):
**Fd dice CAN be turned down using Fire**
source: staging game 463
If using Fire was something other than a skill attack, then pairing it with d on the same die would disallow the Fire turndown.

**Comment by PoshFrosh** [comment]
Additionally, with regards to the priginal rules and their intent, I will add this section from The Buttonmen Book, which I think is relevant to this discussion:

> Design Notes: Jeffrey Watts proposed several variations of the Fire Die before he and James settled on the current rules.
One of the restrictions that was proposed, but was later determined to be moot, was that the Fire Die was disallowed
from participating in the attack it helped to enable. The two designers figured out that it was already impossible for the
die to participate in a Power Attack being undertaken by another die, even ignoring the fact that Fire Dice are specifically
prohibited from making Power Attacks. And, should the user wish to involve the Fire Die in a Skill Attack it had
assisted, he could get the same result by simply including the die in the attack without using its special power!

That last sentence is simply not true, in my opinion:
If you use the F die to contribute to a skill attack in the normal way, it would reroll.
If you use the F die to contribute to a skill attack using Fire turndown, then it will not reroll.
So I'm confused about their logic. (Am I missing something?)

Anyway, what I think is important, is that this paragraph suggests to me **their design intention: they didn't want Fire dice to be able to perform Power or Skill attacks. They wanted them to be used for Fire turndown. So they prevented Power Attacks. And the only reason they allowed skill attacks was because they thought they were somehow "equivalent" to using Fire.**

So, if we assume that intent to be true, how does it affect our discussion of how Fire would interact with konstant? To me it suggest the Fk should be "allowed".
